### PR TITLE
PERF: pass through to numpy validation for iloc setitem

### PR DESCRIPTION
**Pull request by jbrockmendel**
We lose a little bit of ground on the range and slice (not really sure why), pick up a bigger amount of ground on list or ndarray.

```
In [3]: ser = pd.Series(range(10**5))                                                                                      
In [4]: key = range(100, 200) 
In [5]: key2 = list(key)
In [6]: key3 = slice(100, 200)                                                                                             
In [7]: key4 = np.array(key2)                                                                                                 

In [16]: %timeit ser.iloc[key] = 1
56.3 µs ± 1.19 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)  # <-- master
62.9 µs ± 1.77 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)  # <-- PR

In [17]: %timeit ser.iloc[key2] = 1
95.6 µs ± 2.69 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)  # <-- master
55.6 µs ± 922 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)  # <-- PR

In [20]: %timeit ser.iloc[key3] = 1
49 µs ± 756 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)  # <-- master
50.6 µs ± 1.01 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)  # <-- PR

In [21]: %timeit ser.iloc[key4] = 1
71.6 µs ± 1.98 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)  # <-- master
45.7 µs ± 427 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)  # <-- PR
```

### CODE REVIEW

File: pandas/core/indexing.py [hunk 1]
```diff
@@ -1500,18 +1499,14 @@ def _convert_to_indexer(self, key, axis: int, is_setter: bool = False):
...
-            # _validate_indexer call will always raise
-            labels._validate_indexer("positional", key, "iloc")
+        return key
 
-        self._validate_key(key, axis)
+    def _get_setitem_indexer(self, key):
+        if isinstance(key, tuple):
+            if len(key) > self.ndim:
+                # TODO: we could let this fall through if we are OK
+                #  with having numpy raise IndexError("too many indices for array")
```
Comment (jreback) [inline pandas/core/indexing.py]: would be ok i think

File: pandas/tests/frame/indexing/test_indexing.py [hunk 2]
```diff
@@ -27,6 +27,9 @@
 
 from pandas.tseries.offsets import BDay
 
+# We pass through a TypeError raised by numpy
```
Comment (jreback) [inline pandas/tests/frame/indexing/test_indexing.py]: nit. you used iloc_msg in the other function (non-private); i would make these more verbose

slice_exception_msg

**Review by jreback — CHANGES_REQUESTED** [review]
lgtm. minor nit

**Comment by jbrockmendel** [comment]
Removed the tuple length validation, picked up on the order of a microsecond compared to the posts from the OP.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks

**Pull request merged**